sources/cfsignal/src/main.rs

```diff
@@ -0,0 +1,94 @@
+#![deny(rust_2018_idioms, unused_imports)]
+
+/*!
+## Introduction
+
+Cfsignal is similar to [cfn-signal] (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-signal.html).
+
+When creating an Auto Scaling Group, CloudFormation can be configured to wait for the expected number of signals from instances before considering the ASG successfully created. See [CreationPolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-creationpolicy.html) and [UpdatePolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html) for more details.
+
+Cfsignal uses `systemctl is-system-running` to determine whether the boot has succeeded or failed, and sends the corresponding signal to the CloudFormation stack.
+
+## Configuration
+
+Configuration is read from a TOML file, which is generated from Bottlerocket settings:
+* `should_signal`: Whether to check system status and send signal.
+* `stack_name`: Name of the CFN stack to signal.
+* `logical_resource_id`: The logical ID of the AutoScalingGroup resource that you want to signal.
+*/
+
+#![deny(rust_2018_idioms)]
+
+mod cloudformation;
+mod config;
+mod error;
+mod system_check;
+
+use crate::config::Config;
+use crate::error::Result;
+use crate::system_check::SystemCheck;
+use cloudformation::signal_resource;
+use log::LevelFilter;
+use log::{error, info, warn};
+use simplelog::{Config as LogConfig, SimpleLogger};
+use std::fs;
+use std::process;
+
+// We only want to run cfsignal once, at first boot.  Our systemd unit file has a
+// ConditionPathExists that will prevent it from running again if this file exists.
+// We create it after running successfully.
+const MARKER_FILE: &str = "/var/lib/bottlerocket/cfsignal.ran";
+
+/// pub(crate) for testing.
+async fn run() -> Result<()> {
+    SimpleLogger::init(LevelFilter::Info, LogConfig::default())
+        .expect("unable to configure logger");
+
+    // load the cfsignal config file
+    let config = Config::from_file("/etc/cfsignal.toml")?;
+
+    let mut signal_status = "FAILURE";
+    let system_check = Box::new(SystemCheck {});
+    let system_status = system_check.system_running()?;
+    info!(
+        "System status is: {} [{}]",
+        system_status.status, system_status.exit_code
+    );
+
+    // run only if the opt-in flag is set
+    if config.should_signal {
```

Comment:
Tiny nit (unless you feel that the marker file needs to be written even when should_signal is false):

```suggestion
    // run only if the opt-in flag is set
    if !config.should_signal {
        return Ok((())
    }
```

Reply:
I think we just want to run this only once regardless of what `should_signal` is initially. For the use-case of signalling cloudformation stacks, the expectation is that this setting is set in userdata in the ASG's launch template. So it should only take effect once. If it wasn't set on first boot, we should just never waste time running this service.

Reply:
totally agree with @etungsten